cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java

```diff
@@ -96,6 +106,76 @@
   }
 
   /**
+   * Migrate schedules in a given namespace from app metadata store to ProgramScheduleStoreDataset
+   * @param namespaceId the namespace with schedules to be migrated
+   * @param appMetaStore app metadata store with schedules to be migrated
+   * @return the lexicographically largest namespace id String with schedule migration completed
+   */
+  public String migrateFromAppMetadataStore(NamespaceId namespaceId, Store appMetaStore) {
+    String completedNamespace = getMigrationCompleteNamespace();
+    if (completedNamespace != null && completedNamespace.compareTo(namespaceId.toString()) > 0) {
+      return completedNamespace;
+    }
+    for (ApplicationSpecification appSpec : appMetaStore.getAllApplications(namespaceId)) {
+      ApplicationId appId = namespaceId.app(appSpec.getName(), appSpec.getAppVersion());
+      for (ScheduleSpecification scheduleSpec : appSpec.getSchedules().values()) {
+        String scheduleKey = rowKeyForSchedule(appId, scheduleSpec.getSchedule().getName());
+        if (!store.get(new Get(scheduleKey)).isEmpty()) {
+          LOG.debug("Skip migrating schedule '{}' since it already exists.", scheduleSpec.getSchedule().getName());
```

Comment:
we do all migration in one namespace within a single transaction, right? So if it fails within the middle of a namespace, it'll the writes from that transaction won't be visible in the store.